# Experimento numérico \# 1 - Tema: Comportamiento de algoritmo para aproximación SVD cambiando tamaño de matriz

**Responsable:** Miguel Angel Millan Dorado

## 0. Objetivo

Este documento pretende mostar los resultados obtenidos con la implementación realizada, en el contexto del comportamiento del algoritmo para aproximación SVD cambiando el tamaño de matriz. Se usará el mismo método para generación de matriz pseudo-aleatoria con tamañoños de $10^1$, $10^2$ y $10^3$.

**Nota:** El reporte general de resultados obtenidos con las implementaciones realizadas en este proyecto se puede consultar en el siguiente [vínculo](https://github.com/mno-2020-gh-classroom/ex-modulo-3-comp-matricial-svd-czammar/blob/master/results/Reporte_resultados.ipynb)

## 1. Consideraciones

Al respecto de los experimentos numéricos realizados para consolidar el presente reporte, pretenden analizar el comportamiento de la implementación para algoritmo para aproximación SVD ante variaciones en el tamaño de matriz correspondiente.

En este sentido, en particular para cada experimento realizado, se reportan según resulte aplicable:

* los parámetros empleados en la simulaciones, 
* las dimensiones de las matrices y vectores involucrados, así como el prodecimiento pseudo-aleatorio que les dio origen, 
* 1) el tiempo involucrado en correr los experimentos, 
* 2) número de condición de las matrices pseudo-aleatorias, y 
* 3) el error relativo obtenido, para la solución de un sistema lineal de la forma $Ax=b$, es decir el cociente

$$|| Ax - b ||_2 / || b ||_2$$
* 4) La norma de $A$ y de la matriz obtenida al rearmar A tras multiplica las matrices resultantes de la aproximación de la descomposición SVD vía el algoritmo **One-Sided Jacobi**


### 1.1 Consideraciones sobre la infraestructura empleada

**Especificaciones de ambiente común de trabajo**

Para realizar el presente experimento numérico se ha empleado la imagen de docker basada en R del curso MNO 2020 (palmoreck/jupyterlab_r_kernel:1.1.0)

```
docker run --rm -v `pwd`:/datos --name jupyterlab_r_kernel_local -p 8888:8888 -d palmoreck/jupyterlab_r_kernel:1.1.0

```

*Nota:* el comando "-v \`pwd\`:/datos", permite montar el directorio actual en donde el usuario se encuentre situada en terminal como un volumen de la imagen de docker, dentro del directorio "/datos".

## 2. Experimento numérico

**Objetivo**

Observar el comportamiento y tiempos de ejecución del algotirmo para aproximación SVD para diferentes tamaños de su matriz de entrada A, en específico para tamañanos de $10^1$, $10^2$ y $10^3$.

A tal respecto, se destaca que se realizó este experimentos buscando probar que el algoritmo no presente fallos para tamaños diferentes de la matriz de entrada, así como medir los tiempos de ejecución.

**Cargamos codigo desarrollado previamente**

* **utils.R:** contiene las funciones auxiliares desarrolladas para el proyecto
* **00-load.R:** la implementación del método de eliminación por bloques, empleando la aproximación de la descomposición SVD vía el algoritmo **One-Sided Jacobi**.

In [1]:
## Instalamos paquetes
rm(list = ls())

paquetes <- c('matrixcalc')

instalar <- function(paquete) {
  if (!require(paquete,character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)) {
    install.packages(as.character(paquete), dependecies = TRUE, repos = "http://cran.us.r-project.org")
    library(paquete, character.only = TRUE, quietly = TRUE, warn.conflicts = FALSE)
  }
}


lapply(paquetes, instalar)

Warning message in library(package, lib.loc = lib.loc, character.only = TRUE, logical.return = TRUE, :
“there is no package called ‘matrixcalc’”
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



[[1]]
[1] "matrixcalc" "stats"      "graphics"   "grDevices"  "utils"     
[6] "datasets"   "methods"    "base"

In [61]:
## Cargamos paquetes necesarios
library("matrixcalc")

#source("metadata.R")
source("utils.R")
source("00-load.R")

### 2.1 Experimento X.1 Aproximación SVD con matriz de $10^1$

Este experimento se basa en los siguientes parámetros:

| # |         Parámetros         | Valor/Rango de valores | Comentarios |
|:-:|:--------------------------:|:----------------------:|:-----------:|
| 1 |     Dimensiones de $A$     |           $10x10$             |      n=10        |
| 2 |         Tolerancia         |                $10^{-8}$        |             |
| 3 |          Maxsweep          |               20         |             |
| 4 |   Metodo pseudo-aleatorio  |             $rnorm(n^{2})$           |           |

**Codigo:** Algoritmo que aplica el método de Jacobi para hacer la descomposición SVD de una matriz de entrada A (nxn).

In [70]:
set.seed(231)
n= 10**1
A = matrix(rnorm(n**2), ncol=n)
TOL<-10**-8
maxsweep<-20
#Función aproximación SVD
svd<-svd_jacobi_aprox(A,TOL,maxsweep)

In [71]:
system.time(svd<-svd_jacobi_aprox(A,TOL,maxsweep))

   user  system elapsed 
  0.085   0.000   0.085 

In [72]:
#Regresamos a la matriz original usando SVD
A_prima = svd$U%*%svd$S%*%t(svd$V)

In [73]:
#Comprobamos que las matrices seán iguales.
identical(round(A,5),round(A_prima,5))

[1] TRUE

**Resultados**

Los resultados obtenidos se resumen a través de la siguiente tabla:

| # |                Parámetros                | Valor/Rango de valores |            Comentarios            |
|:-:|:----------------------------------------:|:----------------------:|:---------------------------------:|
| 1 |            Tiempo de ejecución           |                 0.078 s       |                                   |
| 2 |      $A = U*S*V^{t}$     |                    True    |  Si regresa a la matriz de entrada A                                 |


### 2.2 Experimento X.1 Aproximación SVD con matriz de $10^2$

Este experimento se basa en los siguientes parámetros:

| # |         Parámetros         | Valor/Rango de valores | Comentarios |
|:-:|:--------------------------:|:----------------------:|:-----------:|
| 1 |     Dimensiones de $A$     |           $100x100$             |      n=100        |
| 2 |         Tolerancia         |                $10^{-8}$        |             |
| 3 |          Maxsweep          |               20         |             |
| 4 |   Metodo pseudo-aleatorio  |             $rnorm(n^{2})$           |           |

**Codigo:** Algoritmo que aplica el método de Jacobi para hacer la descomposición SVD de una matriz de entrada A (nxn).

In [76]:
set.seed(231)
n= 10**2
A_2 = matrix(rnorm(n**2), ncol=n)
TOL<-10**-8
maxsweep<-20
#Función aproximación SVD
svd_2<-svd_jacobi_aprox(A_2,TOL,maxsweep)

In [77]:
system.time(svd_2<-svd_jacobi_aprox(A_2,TOL,maxsweep))

   user  system elapsed 
 10.666   0.013  10.826 

In [78]:
#Regresamos a la matriz original usando SVD
A_prima2 = svd_2$U%*%svd_2$S%*%t(svd_2$V)

In [80]:
#Comprobamos que las matrices seán iguales.
identical(round(A_2,5),round(A_prima2,5))

[1] TRUE

**Resultados**

Los resultados obtenidos se resumen a través de la siguiente tabla:

| # |                Parámetros                | Valor/Rango de valores |            Comentarios            |
|:-:|:----------------------------------------:|:----------------------:|:---------------------------------:|
| 1 |            Tiempo de ejecución           |                 10.826 s       |                                   |
| 2 |      $A = U*S*V^{t}$     |                    True    |  Si regresa a la matriz de entrada A_2                                 |


### 2.3 Experimento X.1 Aproximación SVD con matriz de $10^3$

Este experimento se basa en los siguientes parámetros:

| # |         Parámetros         | Valor/Rango de valores | Comentarios |
|:-:|:--------------------------:|:----------------------:|:-----------:|
| 1 |     Dimensiones de $A$     |           $1000x1000$             |      n=1000        |
| 2 |         Tolerancia         |                $10^{-8}$        |             |
| 3 |          Maxsweep          |               20         |             |
| 4 |   Metodo pseudo-aleatorio  |             $rnorm(n^{2})$           |           |

**Codigo:** Algoritmo que aplica el método de Jacobi para hacer la descomposición SVD de una matriz de entrada A (nxn).

In [81]:
set.seed(231)
n= 10**3
A_3 = matrix(rnorm(n**2), ncol=n)
TOL<-10**-8
maxsweep<-20
#Función aproximación SVD
svd_3<-svd_jacobi_aprox(A_3,TOL,maxsweep)

In [82]:
system.time(svd_3<-svd_jacobi_aprox(A_3,TOL,maxsweep))

    user   system  elapsed 
1975.023    0.093 1986.465 

In [83]:
#Regresamos a la matriz original usando SVD
A_prima3 = svd_3$U%*%svd_3$S%*%t(svd_3$V)

In [84]:
#Comprobamos que las matrices seán iguales.
identical(round(A_3,5),round(A_prima3,5))

[1] TRUE

**Resultados**

Los resultados obtenidos se resumen a través de la siguiente tabla:

| # |                Parámetros                | Valor/Rango de valores |            Comentarios            |
|:-:|:----------------------------------------:|:----------------------:|:---------------------------------:|
| 1 |            Tiempo de ejecución           |                 1986.465 s       |                                   |
| 2 |      $A = U*S*V^{t}$     |                    True    |  Si regresa a la matriz de entrada A_3                                 |


**Resumen de resultados obtenidos en experimentos**

Los resultados obtenidos, se compararan a través de la siguiente:

| # | Experimento | Tiempo de ejecución (s) |     $A = U*S*V^{t}$  | 
|:-:|:-----------:|:-------------------:|----------------|
| 1 |     X.1     |  0.078              |         T      |               
| 2 |     X.2     |       10.826              |    T           |               
| 3 |     X.3     |        1986.465              |  T              |                

[Pendiente: Incluir gráficos de considerarlo relevante para el análisis]

## 3. Principalez hallazgos

Al respecto, se destacan los siguientes hallazgos del experimento:

* Hallazgo 1: Los tiempos de ejecución aumentaron de forma considerable al aumentar la dimensión de la matriz en potencias de 10.
* Hallazgo 2: El Algoritmo fue consistente con estas dimensiones,devolviendo la SVD correspondiente.

